### PR TITLE
Add train test split to DirectoryIterator

### PR DESCRIPTION
**Pull request by ksindi**
When doing transfer learning as described in Keras [applications](https://keras.io/applications/), you have to manually create a validation subdirectory for your validation set. This PR augments `DirectoryIterator` by giving the ability to dynamically assign a percentage of images in the training data as the validation set.

My approach is to hash the file names and do a variant assignment (in a similar vein to tensorflow's transfer learning [script](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/image_retraining/retrain.py)). Because it's based on file names, the hash algorithm won't shuffle files between sets when called again.

The usage looks like this:
```
# instantiate an ImageDataGenerator with a validation percentage
generator = image.ImageDataGenerator(validation_split=0.1)
train_iterator = generator.flow_from_directory(image_dir, subset='training')
validation_iterator = generator.flow_from_directory(image_dir, subset='validation')
model.fit_generator(train_iterator, validation_data=validation_iterator)
```

One downside to this PR is the variant assignment is done twice, both for the train iterator and validation iterator. With the self-imposed requirement of doing as little refactoring as possible, I thought this was okay given that step is relatively fast.

### CODE REVIEW

**Comment by fchollet** [comment]
I'm not convinced this is necessary. With the current iterators, all you have to do is put your validation data in a separate folder.

**Comment by ksindi** [comment]
I think this approach is more general so you don't have to create a validation directory and you can dynamically change the percentage allocated for validation. At least for me it also makes my pipeline for training simpler. Happy to close if you disagree.

**Comment by Dref360** [comment]
Having a 3rd folder for validation is quite restrictive and for most of the datasets out there, this means that we need to create the folder ourselves.

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -933,11 +969,23 @@ def _recursive_list(subpath):
                             is_valid = True
                             break
                     if is_valid:
-                        self.classes[i] = self.class_indices[subdir]
-                        i += 1
-                        # add filename relative to directory
+                        if do_train_validation_split:
+                            # perform variant assignment and determine whether to ignore files from list
+                            hash_pct = _calc_hash_pct(fname)
+                            valid_validation = hash_pct < validation_pct and not is_training
```
Comment (Dref360) [inline keras/preprocessing/image.py]: Instead of hashing, why not just generate random numbers? Do we really care about the randomness of the valid set?

Comment (ksindi) [inline keras/preprocessing/image.py]: Using hashing is a nice way to maintain the same files in the their respective sets even if more files are added.

**Comment by Christina-Young** [comment]
This change would be very useful.  It is a real hassle to have to create separate validation directories, especially when you want to use "on the fly" cross validation strategies such as stratified k-fold with flow_from_directory(), which would be just one step further than a dynamic train/test split.

**Comment by ksindi** [comment]
@aupilot I essentially referenced tensorflow for `as_bytes` to allow for interoperability. It should work the same: https://github.com/tensorflow/tensorflow/blob/0a4f5b6bda405814af59803829216762e030728d/tensorflow/python/util/compat.py#L46. Should figure out how to credit tensorflow.

**Comment by fchollet** [comment]
We can add this feature, but the API will need to be revised. We use `validation_split` everywhere else.

Can a contributor review this PR?

**Comment by lunardog** [comment]
@ksindi please update the API (`validation_pct`->`validation_split`) and resolve conflicts. It's a much needed feature.

**Comment by ksindi** [comment]
Thanks @lunardog will update the PR. Wasn't sure if I had to sign off from a Keras contributor.

**Comment by ksindi** [comment]
@lunardog made changes. Waiting on tests.

**Comment by ksindi** [comment]
Updated API. All tests pass except for one in `KERAS_BACKEND=cntk` in py 2.7, which seems to be unrelated. cc @fchollet @lunardog

**Comment by ksindi** [comment]
All tests passing @lunardog cc @fchollet

**Comment by venuktan** [comment]
@fchollet @ksindi whats the status on this ?

**Comment by ksindi** [comment]
@venuktan I'm done on my end. Just waiting for review.

**Comment by danielmorvay** [comment]
This feature (Add train test split to DirectoryIterator) could be important indeed.
Let's assume, you have 10 million pics, you split the train, valid, dev sets into 3 sets, i,e: 80%,10%,10% (test is separated also, on top of the 100%) then you decide to change the split to 95%,5%,5%. You might need to delete the whole 1M pics, and start from scratch.

**Comment by mishafain** [comment]
I think this change would be quite useful. I've pulled the code to one of my projects and it did make a difference - it was as straightforward as using train_test_split or similar utilities, but instead with the image data generator. Otherwise I'd have to run re-run the script that splits the image folder into train and validation folder every time I change the split or add extra data (which was coming in continuously).

**Comment by ksindi** [comment]
@fchollet please let me know if there's anything I need to do on my end

**Comment by fchollet** [comment]
@ksindi sorry for the delay in reviewing. I will look at it asap (or if another contributor would like to review, that would be great too). I still think this is a good feature to add.

**Comment by fchollet** [comment]
I've looked at the PR. I believe we should implement this feature differently.

1 - The overall API looks good to me. However it should also be supported in `flow`.

```python
generator = image.ImageDataGenerator(validation_split=0.5)
train_iterator = generator.flow_from_directory(image_dir, subset='training')
val_iterator = generator.flow_from_directory(image_dir, subset='validation')

# This should also be possible
train_iterator = generator.flow(x, y, subset='training')
val_iterator = generator.flow(x, y, subset='validation')
```

2 - This should be implemented by returning *different* generators for each mode, in `def flow()` and `def flow_from_directory()`.

3 - For `flow` this is trivial as you can just split the arrays and instantiate a generator for the corresponding split (e.g. `x[split:]` and `y[split:]`).

4 - For `flow_from_directory`, we can pass a constructor argument to `DirectoryIterator` that says "only take into account the first (or last) x% files in each directory". The argument would then be passed to `_count_valid_files_in_directory` and `_list_valid_filenames_in_directory`.

I would recommend naming it `split` and having it be a tuple of two floats, e.g. `split=(0.6, 1)` for "last 40%" and `split=(0, 0.2)` for "first 20%".

It would only be a few lines of changes overall. Does it make sense?

**Comment by ksindi** [comment]
Thanks for the feedback @fchollet. Will make changes.

**Comment by lunardog** [comment]
> 4 - For flow_from_directory, we can pass a constructor argument to DirectoryIterator that says "only take into account the first (or last) x% files in each directory". 

In this implementation, how would we go about selecting a *random* subset of x% files for validation? Applying `shuffle=True` to each `flow` or `flow_from_directory` should shuffle the sample order in each subset. How about another parameter for shuffling before splitting: 

```
generator = image.ImageDataGenerator(validation_split=0.5, random_split=True)
```

If `random_split` is `False` it would take the first/last 50%. If it's `True`, it should select a random subset of 50% items as test and the rest as validation.

**Comment by fchollet** [comment]
@lunardog this is by design: in `fit`, `validation_split` will select as validation set the *last* N samples, *before* shuffling. This means that your validation set is reproducible and controllable.

We want the same behavior here, for consistency. So we select the last x% images in each folder.

**Comment by lunardog** [comment]
Thank you for clarifying @fchollet . Good points about reproducibility and consistency.

File: keras/preprocessing/image.py [hunk 2]
```diff
@@ -991,22 +1036,36 @@ def _list_valid_filenames_in_directory(directory, white_list_formats,
...
+    is_training = subset is not None and subset == 'training'
+    dirname = os.path.basename(directory)
+
     classes = []
     filenames = []
-    subdir = os.path.basename(directory)
-    basedir = os.path.dirname(directory)
     for root, _, files in _recursive_list(directory):
-        for fname in sorted(files):
+        for fname in files:
```
Comment (fchollet) [inline keras/preprocessing/image.py]: You can make the whole PR a lot simpler by using: `for fname in sorted(files)[start : stop]`

Remember: this feature should only total a few lines of change. The only "logic" required by this PR is the computation of `start` and `stop` above given the `validation_split` passed by the user (if any).

Comment (ksindi) [reply keras/preprocessing/image.py]: @fchollet just for clarification, I guess you're saying I should do away with the filename hashing for the split and just use `split=(start, stop)` ? My current implementation of filename hashing already makes the validation split consistent without having to specify the segment.

Also would `stop` be necessary or can we just treat the `validation_split` as the length? E.g. `validation_split=0.3` and `split_start=0.7` would mean treat last 30% of each directory as validation.

Comment (fchollet) [reply keras/preprocessing/image.py]: `start`/`stop` is flexible since it means `_count/list_valid_files_in_directory` will be agnostic to the validation/training story, and only needs the `split` argument. But it doesn't make a big difference.

Comment (ksindi) [reply keras/preprocessing/image.py]: Alright, I'll get rid of the filename hashing in favor of `(start, stop)`. So would the API would look like:
```
generator = image.ImageDataGenerator()
train_iterator = generator.flow_from_directory(image_dir, subset='training', split=(0.0, 0.4))
val_iterator = generator.flow_from_directory(image_dir, subset='validation', split=(0.4, 1.0))

train_iterator = generator.flow(x, y, subset='training', split=(0.0, 0.4))
val_iterator = generator.flow(x, y, subset='validation', split=(0.4, 1.0))
```

Comment (fchollet) [reply keras/preprocessing/image.py]: `split` can be reserved for internal methods, as it isn't very user-friendly. For user-facing APIs, we can go with your original proposal:

```python
generator = image.ImageDataGenerator(validation_split=0.5)
train_iterator = generator.flow_from_directory(image_dir, subset='training')
val_iterator = generator.flow_from_directory(image_dir, subset='validation')

# This should also be possible
train_iterator = generator.flow(x, y, subset='training')
val_iterator = generator.flow(x, y, subset='validation')
```

Comment (ksindi) [reply keras/preprocessing/image.py]: Thanks. Hope to fix by end of this week.

Comment (peachthiefmedia) [reply keras/preprocessing/image.py]: The start/stop usage is quite useful in that you can do cross fold validation very easily with it, but I agree that it isn't very user-friendly. If it was still on the user api side it would be nice though.
Also thanks @ksindi for this update, I have also used it and it makes a lot of things easier.

**Comment by ksindi** [comment]
@fchollet I refactored to using split internally, as discussed.
```
generator = image.ImageDataGenerator(validation_split=0.5)
train_iterator = generator.flow_from_directory(image_dir, subset='training')
val_iterator = generator.flow_from_directory(image_dir, subset='validation')

# This should also be possible
train_iterator = generator.flow(x, y, subset='training')
val_iterator = generator.flow(x, y, subset='validation')
```

Let me know what you think.

File: keras/preprocessing/image.py [hunk 3]
```diff
@@ -438,7 +438,8 @@ def __init__(self,
                  vertical_flip=False,
                  rescale=None,
                  preprocessing_function=None,
-                 data_format=None):
+                 data_format=None,
+                 validation_split=None):
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Set the default value to `0.0` rather than None, for consistency with `fit`

File: keras/preprocessing/image.py [hunk 4]
```diff
@@ -473,6 +474,10 @@ def __init__(self,
             self.channel_axis = 3
             self.row_axis = 1
             self.col_axis = 2
+        if validation_split is not None and not 0 <= validation_split <= 1:
```
Comment (fchollet) [inline keras/preprocessing/image.py]: `not 0 < validation_split < 1` (0 and 1 are invalid values)

File: keras/preprocessing/image.py [hunk 5]
```diff
@@ -865,22 +873,45 @@ class NumpyArrayIterator(Iterator):
...
                              (np.asarray(x).shape, np.asarray(y).shape))
-
+        validation_split = image_data_generator._validation_split
+        if subset is not None:
+            # check if self.image_data_generator.validation_split is None
+            if validation_split is None:
+                raise ValueError('Creating a subset with validation_split '
+                                 'set as None in NumpyArrayIterator',
+                                 validation_split)
+            if not 0 <= validation_split <= 1:
```
Comment (fchollet) [inline keras/preprocessing/image.py]: This is already validated in the constructor, no need.

File: keras/preprocessing/image.py (hunk 4 above)
Comment (ksindi) [reply keras/preprocessing/image.py]: @fchollet wouldn't it fail if `validation_split` defaults to 0.0, as suggested above?

**Comment by fchollet** [comment]
Replace "if validation_split is not None" with "if validation_split"

**Comment by fchollet** [comment]
@Dref360 any comment on this, or should I merge?

**Comment by ksindi** [comment]
thanks @fchollet @Dref360

**Comment by chrisrb10** [comment]
Hi all,
This function is exactly what I need for the project I am working on - to use exactly the use case @ksindi originally outlined creating a training and validation set from a single directory of images (in class sub-directories).
Is this functionality available in a release version of Keras?
Thanks

**Comment by ksindi** [comment]
Thanks @chrisrb10. It just got merged so unlikely. You can track releases here: https://github.com/keras-team/keras/releases/.
